File: system-design/oop-patterns/04-factory-method-pattern/final/cpp/main.cpp
// Factory Method Pattern - final
//
// RandomAnimalFactory and BalancedAnimalFactory both implement AnimalFactory,
// so populateZoo can accept either one without ever mentioning Dog, Cat,
// Duck, or either factory by name.

#include <iostream>
#include <string>
#include <memory>
#include <cstdlib>

class Animal {
public:
    virtual ~Animal() = default;
    virtual std::string name() const = 0;
};

class Dog : public Animal {
public:
    std::string name() const override { return "Dog"; }
};

class Cat : public Animal {
public:
    std::string name() const override { return "Cat"; }
};

class Duck : public Animal {
public:
    std::string name() const override { return "Duck"; }
};

class AnimalFactory {
public:
    virtual ~AnimalFactory() = default;
    virtual std::unique_ptr<Animal> createAnimal() = 0;
};

class RandomAnimalFactory : public AnimalFactory {
public:
    std::unique_ptr<Animal> createAnimal() override {
        int choice = std::rand() % 3;
        if (choice == 0) return std::make_unique<Dog>();
        if (choice == 1) return std::make_unique<Cat>();
        return std::make_unique<Duck>();
    }
};

class BalancedAnimalFactory : public AnimalFactory {
public:
    std::unique_ptr<Animal> createAnimal() override {
        if (dogCount_ <= catCount_ && dogCount_ <= duckCount_) {
            dogCount_++;
            return std::make_unique<Dog>();
        }
        if (catCount_ <= duckCount_) {
            catCount_++;
            return std::make_unique<Cat>();
        }
        duckCount_++;
        return std::make_unique<Duck>();
    }

private:
    int dogCount_ = 0;
    int catCount_ = 0;
    int duckCount_ = 0;
};

void populateZoo(AnimalFactory& factory, int count) {
    for (int i = 0; i < count; ++i) {
        auto animal = factory.createAnimal();
        std::cout << animal->name() << "\n";
    }
}

int main() {
    std::cout << "Random:\n";
    RandomAnimalFactory randomFactory;
    populateZoo(randomFactory, 6);

    std::cout << "Balanced:\n";
    BalancedAnimalFactory balancedFactory;
    populateZoo(balancedFactory, 6);

    return 0;
}

File: system-design/oop-patterns/04-factory-method-pattern/final/ts/main.ts
// Factory Method Pattern - final
//
// RandomAnimalFactory and BalancedAnimalFactory both implement AnimalFactory,
// so populateZoo can accept either one without ever mentioning Dog, Cat,
// Duck, or either factory by name.

interface Animal {
  name(): string;
}

class Dog implements Animal {
  name(): string { return 'Dog'; }
}

class Cat implements Animal {
  name(): string { return 'Cat'; }
}

class Duck implements Animal {
  name(): string { return 'Duck'; }
}

interface AnimalFactory {
  createAnimal(): Animal;
}

class RandomAnimalFactory implements AnimalFactory {
  createAnimal(): Animal {
    const choice = Math.floor(Math.random() * 3);
    if (choice === 0) return new Dog();
    if (choice === 1) return new Cat();
    return new Duck();
  }
}

class BalancedAnimalFactory implements AnimalFactory {
  private dogCount = 0;
  private catCount = 0;
  private duckCount = 0;

  createAnimal(): Animal {
    if (this.dogCount <= this.catCount && this.dogCount <= this.duckCount) {
      this.dogCount++;
      return new Dog();
    }
    if (this.catCount <= this.duckCount) {
      this.catCount++;
      return new Cat();
    }
    this.duckCount++;
    return new Duck();
  }
}

function populateZoo(factory: AnimalFactory, count: number): void {
  for (let i = 0; i < count; i++) {
    console.log(factory.createAnimal().name());
  }
}

console.log('Random:');
populateZoo(new RandomAnimalFactory(), 6);

console.log('Balanced:');
populateZoo(new BalancedAnimalFactory(), 6);
